commit/merkleroot/rmn/peerclient.go

```diff
@@ -31,27 +52,150 @@ type PeerResponse struct {
 }
 
 type peerClient struct {
-	ragep2ptypes.Address
-	peer       ragep2ptypes.PeerInfo
-	rmnPeerIDs mapset.Set[rmntypes.NodeID]
-	respChan   chan PeerResponse
+	lggr                        logger.Logger
+	peerGroupFactory            networking.PeerGroupFactory
+	respChan                    chan PeerResponse
+	peerGroup                   networking.PeerGroup // nil initially, until InitConnection is called
+	genericEndpointConfigDigest cciptypes.Bytes32
+	rageP2PStreams              map[rmntypes.NodeID]networking.Stream
+	bootstrappers               []commontypes.BootstrapperLocator
+	mu                          *sync.RWMutex
 }
 
-func NewPeerClient(peer ragep2ptypes.PeerInfo, rmnNodeIDs []rmntypes.NodeID) PeerClient {
+func NewPeerClient(
+	lggr logger.Logger,
+	peerGroupFactory networking.PeerGroupFactory,
+	bootstrappers []commontypes.BootstrapperLocator,
+) PeerClient {
 	return &peerClient{
-		peer:       peer,
-		rmnPeerIDs: mapset.NewSet(rmnNodeIDs...),
-		respChan:   make(chan PeerResponse),
+		lggr:                        lggr,
+		peerGroupFactory:            peerGroupFactory,
+		respChan:                    make(chan PeerResponse),
+		peerGroup:                   nil,
+		rageP2PStreams:              make(map[rmntypes.NodeID]networking.Stream),
+		genericEndpointConfigDigest: cciptypes.Bytes32{},
+		bootstrappers:               bootstrappers,
+		mu:                          &sync.RWMutex{},
+	}
+}
+
+func (r *peerClient) InitConnection(
+	_ context.Context,
+	commitConfigDigest, rmnHomeConfigDigest cciptypes.Bytes32,
+	peerIDs []string,
+) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	h := sha256.Sum256(append(commitConfigDigest[:], rmnHomeConfigDigest[:]...))
+	r.genericEndpointConfigDigest = writePrefix(ocr2types.ConfigDigestPrefixCCIPMultiRoleRMNCombo, h)
+
+	peerGroup, err := r.peerGroupFactory.NewPeerGroup(
+		[32]byte(r.genericEndpointConfigDigest),
+		peerIDs,
+		r.bootstrappers,
+	)
+
+	if err != nil {
+		return fmt.Errorf("new peer group: %w", err)
+	}
+
+	r.peerGroup = peerGroup
+	return nil
+}
+
+func (r *peerClient) Close() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// individual streams are closed by the peer group
+	if err := r.peerGroup.Close(); err != nil {
+		return fmt.Errorf("close peer group: %w", err)
+	}
+
+	return nil
+}
+
+func (r *peerClient) Send(rmnNode rmntypes.HomeNodeInfo, request []byte) error {
+	if r.peerGroup == nil {
+		return ErrNoConn
+	}
+
+	stream, err := r.getOrCreateRageP2PStream(rmnNode)
```

Comment:
I think this code is a bit racey, it seems cleaner to me to grab the mutex on line 123 and then hold it for the entirety of fetching the stream + sending the message. The reasoning here is that in between this call returning and line 129, the stream could technically get closed?